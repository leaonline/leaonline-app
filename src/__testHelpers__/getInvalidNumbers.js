export const getInvalidNumbers = () => [
  null,
  undefined,
  '1',
  Number.MAX_VALUE * 2,
  -Number.MAX_VALUE * 2,
  {},
  NaN,
  Infinity,
  () => {},
  true,
  false
]