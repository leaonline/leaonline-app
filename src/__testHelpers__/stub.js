import sinon from 'sinon'

export const stub = (target, prop, fn) => {
  const stub = sinon.stub(target, prop)
  stub.callsFake(fn)
  return stub
}

export const restoreAll = () => sinon.restore()
