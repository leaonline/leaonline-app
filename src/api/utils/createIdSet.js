/**
 * Creates a set of all values for a list of documents and respective fieldnames to look for.
 * @param docs {object[]}
 * @param fieldNames {string|string[]}
 * @param includeNull {boolean} flag to include null values in map
 * @return {Set<any>}
 */
export const createIdSet = (docs, fieldNames, includeNull = false) => {
  const ids = new Set()
  const fields = Array.isArray(fieldNames)
    ? fieldNames
    : [fieldNames]
  docs.forEach(doc => {
    fields.forEach(fieldName => {
      const value = doc[fieldName]
      const isDefined = value !== undefined
      const include = value !== null || includeNull
      if (isDefined && include) {
        ids.add(value)
      }
    })
  })
  return ids
}
