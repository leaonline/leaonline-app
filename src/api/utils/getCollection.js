import { Mongo } from 'meteor/mongo'

/**
 * Access (persistent; non-local; named) Mongo.Collection by name
 * @category api
 * @module getCollection
 * @param name {string}
 * @return {Mongo.Collection|undefined}
 */
export const getCollection = name => Mongo.Collection.get(name)
