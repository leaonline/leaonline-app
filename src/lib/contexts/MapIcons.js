import React from 'react'
import { Colors } from '../constants/Colors'
import { createContextStorage } from './createContextStorage'

export const MapIcons = {
  name: 'mapIcons'
}

MapIcons.collection = () => {
  throw new Error('MapIcons is not intialized')
}

MapIcons.storage = createContextStorage(MapIcons)

MapIcons.init = async () => {
  return MapIcons.storage.loadIntoCollection()
}

const icons = []

MapIcons.size = () => icons.length

MapIcons.register = component => {
  icons.push(component)
}

let count = 0

MapIcons.getIncrementalIconIndex = () => {
  if (icons.length === 0) {
    return -1
  }

  if (count > icons.length - 1) {
    count = 0
  }

  return count++
}

MapIcons.render = (index) => {
  if (index > icons.length - 1 || index < 0) {
    return null
  }

  const Component = icons[index]
  return (
    <Component
      stroke={Colors.gray}
      fill={Colors.light}
      width={60}
      height={60}
    />
  )
}
