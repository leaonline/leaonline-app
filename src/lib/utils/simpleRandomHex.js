/**
 * Generates a simple random hex string.
 * Uses Math.random so use it only for simple things.
 * @param {number } [len=6]
 * @return {string}
 */
export const simpleRandomHex = (len = 6) => {
  const n = Math.random()
  const s = String(n.toFixed(len)).split('.')[1]
  return Number(s).toString(16)
}
