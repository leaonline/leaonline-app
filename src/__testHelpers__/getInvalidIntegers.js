import { getInvalidNumbers } from './getInvalidNumbers'

export const getInvalidIntegers = () => getInvalidNumbers().concat([
  -1.1,
  0.2 + 0.1,
  Number.MAX_SAFE_INTEGER + 1,
  -Number.MAX_SAFE_INTEGER - 1,
])