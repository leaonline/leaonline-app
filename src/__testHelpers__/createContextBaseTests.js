import { collectionExists, getCollection } from '../lib/infrastructure/collections/collections'
import { createCollection } from '../lib/infrastructure/createCollection'
import { simpleRandomHex } from '../lib/utils/simpleRandomHex'
import AsyncStorage from '@react-native-async-storage/async-storage/jest/async-storage-mock'

export const createContextBaseTests = ({ ctx, custom }) => {
  const storageKey = `contexts-${ctx.name}`

  beforeAll(() => {
    jest.useFakeTimers({ advanceTimers: true })
  })

  if (ctx.collection) {
    describe(ctx.collection.name, () => {
      beforeAll(() => {
        if (!collectionExists(ctx.name)) {
          createCollection({
            name: ctx.name,
            isLocal: true
          })
        }
      })

      afterAll(() => {
        const collection = getCollection(ctx.name)
        ctx.collection = () => collection
      })

      it('throws if collection is not initiated', () => {
        expect(() => ctx.collection())
          .toThrow(`Collection ${ctx.name} not initialized`)
      })
    })
  }

  if (ctx.init) {
    describe(ctx.init.name, function () {
      it('loads sync docs into collection', async () => {
        const _id = simpleRandomHex()
        const doc = { _id, title: 'moo' }
        await AsyncStorage.setItem(storageKey, JSON.stringify(doc))
        await ctx.init()
        const collection = getCollection(ctx.name)
        expect(collection.findOne({ _id }))
          .toStrictEqual({ _version: 1, ...doc })
      })
    })
  }
  if (custom) {
    custom()
  }
}