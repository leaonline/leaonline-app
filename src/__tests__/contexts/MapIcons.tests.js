import React from 'react'
import { MapIcons } from '../../lib/contexts/MapIcons'
import { createContextBaseTests } from '../../__testHelpers__/createContextBaseTests'
import { simpleRandomHex } from '../../lib/utils/simpleRandomHex'
import renderer from 'react-test-renderer';

describe(MapIcons.name, () => {
  createContextBaseTests({ ctx: MapIcons })

  describe(MapIcons.getIncrementalIconIndex.name, () => {
    it('always returns -1 if no icons are registered', () => {
      MapIcons.setField(simpleRandomHex())
      expect(MapIcons.getIncrementalIconIndex()).toBe(-1)
    })
    it('returns a count that rotates around the icons set size', async () => {
      const fieldId = simpleRandomHex()
      await MapIcons.collection().insert({
        fieldId,
        icons: ['foo', 'bar', 'baz']
      })

      MapIcons.setField(fieldId)

      const size = MapIcons.size()
      expect(size).toBe(3)

      let prev = -1
      for (let i = 0; i < size; i++) {
        const count = MapIcons.getIncrementalIconIndex()
        expect(count).toBeGreaterThan(prev)
        prev = count
      }
      // expect reset
      expect(MapIcons.getIncrementalIconIndex()).toBe(0)
    })
  })
  describe(MapIcons.render.name, () => {
    it('returns null if the index is not within bounds', async () => {
      const fieldId = simpleRandomHex()
      const icons = ['edit', 'check', 'times']
      await MapIcons.collection().insert({ fieldId, icons })
      ;[-3, -2, -1, 3, 4, 5].forEach(index => {
        expect(MapIcons.render(index))
          .toBe(null)
      })
    })
    it('renders an icon by given index', async () => {
      const fieldId = simpleRandomHex()
      const icons = ['edit', 'check', 'times']
      await MapIcons.collection().insert({ fieldId, icons })

      MapIcons.setField(fieldId)
      const size = MapIcons.size()

      for (let i = 0; i < size; i++) {
        const MapIcon = MapIcons.render(i)
        const component = renderer.create(MapIcon);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      }
    })
  })
})
