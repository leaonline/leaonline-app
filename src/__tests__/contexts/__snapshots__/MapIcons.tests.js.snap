// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapIcons  renders an icon by given index 1`] = `
<View
  style={
    {
      "overflow": "hidden",
    }
  }
>
  <View>
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "justifyContent": "center",
        }
      }
    >
      <Text />
    </View>
  </View>
</View>
`;

exports[`mapIcons  renders an icon by given index 2`] = `
<View
  style={
    {
      "overflow": "hidden",
    }
  }
>
  <View>
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "justifyContent": "center",
        }
      }
    >
      <Text />
    </View>
  </View>
</View>
`;

exports[`mapIcons  renders an icon by given index 3`] = `
<View
  style={
    {
      "overflow": "hidden",
    }
  }
>
  <View>
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "justifyContent": "center",
        }
      }
    >
      <Text />
    </View>
  </View>
</View>
`;
